pylint/test/functional/no_else_return.py

```diff
@@ -24,7 +24,7 @@ def foo2(x, y, w, z):
 
 
 def foo3(x, y, z):
-    if x:  # [no-else-return]
```

Comment:
I think this should be flagged as well, given that both branches from this condition are returning, so after rewriting it, it should be:
```
if x:
   if y:
      return y
   return x
return z
```

Reply:
I think I understand how to detect the case where an if statement has a nested if/else that always returns by looking at all If-node children and checking that all paths return.

But in thinking about it some more, I don't think we should do it for this no-else-return. Adding this check is a very specific error check that is ultimately unnecessary since, eventually, we end up with the optimized code (as indicated earlier). Also, that's only one of many cases where an `if` body is always returning due to a nested statement. For example, there are three different statements in the body of the following if-statement that result in always returning:

```
def example_func(f, x, y, z):
    if x == 24:
        # This 'if' statement always returns
        if y: 
            return y
        else:
            return x

        # The 'try' statement always returns
        try:
            f(x)
            return x
        except ValueError:
            return z

        # The while statement always returns
        while True:
            return y

    else:
        return z
    return x
```

@PCManticore  - what do you think?

Reply:
Yeah, that make sense, I think it's probably good as it is right now in your PR.